### ISSUE TITLE
Adding many features

### ISSUE THREAD

**Issue by nullchilly** [xeluxee/competitest.nvim#6]
I will gladly pull request these things but I thought I would let you know if you are working on something similar and your thought on this.

Note: This is written in a hurry, will be editing...

# Split mode
![image](https://user-images.githubusercontent.com/56817415/173244565-bd5341e5-c070-4749-abe5-6dbbbebfd2ac.png)

Will consist of 6 windows: complication error, stdin, stderr, stdout, answer. Will be fully customizable. Personally I don't need complication error tab because theere is already lsp

E.x. `{"set nosplitright | vs | setl wfw | wincmd w | bel sp | vs | vs | 1wincmd w", {1, 2, 3, 4, 5}}`

# Get the most out of companion
Use regrex to parse problem name, let user determine the path to it.

E.x. 
```lua
"https://codeforces.com/contest/(%d+)/problem/(%d)" -> vim.loop.os_homedir() .. "/code/contest/codeforces/$1/$2"
{"https://atcoder.jp/contests/(%s+)/tasks/(%s+)" -> vim.loop.os_homedir() .. "/code/contest/atcoder/$1/$2"
```

# Keybinds
Add keybind tips: quick test case managing, for example: 
- 2t switch to test 2
- <keybind> hide current test, hide all test, hide AC tests, hide WA test

# Auto parsing
Always listen to companion port since we are parsing new problem instead of just to the current file

# Adding user custom output
g++ sol.cpp < 1.in > 2.out 2> 2.err
This will greatly increase

# Change C++ flags on the fly
```
g++ -Wall -O3
g++ -g -O2
```
Maybe compile_commands will helps?

# Saving 
Option to remember state of the current file, like submission history, local flags, test cases status

# neovim style commands
```
:Competitest run 1
:Competitest submit
```
It makes auto complete much easier

# nvim-dap
Debugging is very essential, we need to intergrate with it

# stresstest
Also very essential, I will go with codeforces's library

# And more, will be updating after 15/6/22

**Comment by xeluxee** [comment]
Hi @nullchilly, thanks for your suggestions. Unfortunately I can't start working on these features immediately, I'll start improving CompetiTest in some weeks.

> Split mode

Already [planned](https://github.com/xeluxee/competitest.nvim#roadmap).

> Get the most out of companion

Would you like competitest to clone an entire contest or problem? Related #5

> Auto parsing - Always listen to companion port since we are parsing new problem instead of just to the current file

What's the issue? You can launch `:CompetiTestReceive` multiple times with multiple files

> Adding user custom output and local problem

I can't get your point.
- What do you mean with local problem?
- Aren't testcases (with input and output) enough?
- Do you want to save output on a file?

> Change C++ flags on the fly

Actually you can do it: `:lua require('competitest').setup { compile_command = { cpp = { args = {"$(FNAME)", "-o", "$(FNOEXT)"}}}}`. This isn't so straightforward, but you can make a custom command to do that.
Also `compile_commands.json` is generated by CMake, why one should use CMake for a competitive programming task?
Maybe a local configuration (`.competitest.lua`) fits better for you. See https://github.com/xeluxee/competitest.nvim#local-configuration

> Saving

Saving only for current session or after quitting Neovim?

> neovim style commands

Looks good, I'll think about it

> Debugging

What do you mean exactly? Isn't `nvim-dap` enough? Maybe you'd like to debug a solution with a specific input among the available ones?

> Stress Testing

What is stress testing?

**Comment by nullchilly** [comment]
I am on phone so I will leave a quick reply

I initially wanted to continue update my own plugin since it contains a lot of opinionated but I think it would be better if I can work with someone more familiar with lua to follow good practice

Local problems needs input file like freopen("r", "input.txt"). As you might notice there is also IOI format, OI format not just regular standard IO

About debugging yes you take input then debug directly, you also have to use a different flags just for debugging since it takes longer to compile
It is also better to just use compile_commands since a lot of user are more familiar with it.

Stress test is when you randomly generate some test that works for brute forces solution but doesnt for your current code so you can actually debug them.

_Unfortunately I can't start working on these features immediately_
My intention isn't to ask you to implement all of these but I will do it. You will se why all of these features on my fork. If it doesn't disappoint you I will start pull requests :)

**Comment by nullchilly** [comment]
@xeluxee Better late than never

After spending an entire day editing the source I found myself wanting to change everything in this project

For example: I don't need singlefile testcases, I don't need a floating window that covers the code. 
I want to have per tab config not per buffer. I want output file for checker support. I want to limit input size. Adding split support requires too much refactoring so on and so forth. The times that I spend will be much, much more.

**tl;dr** I have decided to continue developing my [old plugin](https://github.com/nullchilly/cp.nvim) (I will credit you of course ;)
Definitely add these features back if you want. Or stay on the lightweight side 🤔

**Comment by xeluxee** [comment]
I'm sorry you could not contribute to this project. I've started CompetiTest mainly to meet my needs during cp contests, but I'm also interested in hearing other users' needs in order to make this plugin better. The features you've proposed look good, probably I'll implement them in the (near) future.
Anyway good luck with your plugin
Leaving this issue open since other users may like to give suggestions about the mentioned features

**Comment by Tsuzat** [comment]
Hey. I love this project man. Just wanted to know if there is any thing we can do to use just one file. May be JSON or Something for storage. Making 2 files for each test case seems overkill.

**Comment by xeluxee** [comment]
> Hey. I love this project man. Just wanted to know if there is any thing we can do to use just one file. May be JSON or Something for storage. Making 2 files for each test case seems overkill.

https://github.com/xeluxee/competitest.nvim#storing-testcases-in-a-single-file

**Comment by Tsuzat** [comment]
Just looked into doc and found it. By the way thanks for quick reply.
So another question is can we make it transparent? I mean the floating window that shows the test cases and all. Can we remove background from it?

**Comment by Tsuzat** [comment]
> > So another question is can we make it transparent? I mean the floating window that shows the test cases and all. Can we remove background from it?
> 
> At the moment it isn't possible, but I'll add an option to customize window highlights

Thanks for your time.

**Comment by Tsuzat** [comment]
Got it working. Had to do some research  and tuns out it uses NormalFloat
```vim
hi NormalFloat ctermbg=BLACK ctermfg=NONE guibg=NONE guifg=NONE
```
![Screenshot_6](https://user-images.githubusercontent.com/70003855/178017884-9e6d64ed-1dbd-42af-ad4b-c23bf8e93964.png)

**Comment by nullchilly** [comment]
@Tsuzat Maybe you can add a transparent option so you can actually see the code underneath?

https://twitter.com/neovim/status/1145446583573594112

**Comment by xeluxee** [comment]
Split mode has just been implemented in c86a94c5bd2a1edb69d2f77616005423a65949e0
Also take a look at [customizable layouts](https://github.com/xeluxee/competitest.nvim#customize-ui-layout) (idea  taken from @nullchilly's cpeditor.nvim)

**Comment by xeluxee** [comment]
Now you can download problems and contests  with competitive-companion
See https://github.com/xeluxee/competitest.nvim#receive-testcases-problems-and-contests

**Comment by xeluxee** [comment]
> # Get the most out of companion
> 
> Use regrex to parse problem name, let user determine the path to it.
> 
> E.x.
> 
> ```lua
> "https://codeforces.com/contest/(%d+)/problem/(%d)" -> vim.loop.os_homedir() .. "/code/contest/codeforces/$1/$2"
> {"https://atcoder.jp/contests/(%s+)/tasks/(%s+)" -> vim.loop.os_homedir() .. "/code/contest/atcoder/$1/$2"
> ```

See how to [customize folder structure](https://github.com/xeluxee/competitest.nvim/tree/master#customize-folder-structure) for received problems and contests.

**Comment by xeluxee** [comment]
> # neovim style commands
> 
> ```
> :Competitest run 1
> :Competitest submit
> ```
> 
> It makes auto complete much easier
> 
> See #7

**Breaking change**: Neovim-style single command has just been introduced with 254a81b

**Comment by KorigamiK** [comment]
> Maybe compile_commands will helps?
>> Also compile_commands.json is generated by CMake, why one should use CMake for a competitive programming task?

@xeluxee I think he means the use of `compile_flags.txt` which clangd searches for to provide the configuration.

**Comment by xeluxee** [comment]
@KorigamiK if you want to change compile commands you can create a file called `.competitest.lua` (see [local configuration](https://github.com/xeluxee/competitest.nvim#local-configuration)) and configure [compile and run commands](https://github.com/xeluxee/competitest.nvim#customize-compile-and-run-commands) with it.

> > Maybe compile_commands will helps?
> > > Also compile_commands.json is generated by CMake, why one should use CMake for a competitive programming task?
> 
> @xeluxee I think he means the use of `compile_flags.txt` which clangd searches for to provide the configuration.

Currently there's no plan to support language-specific configuration files, [local configuration](https://github.com/xeluxee/competitest.nvim#local-configuration) should be enough.
